[CMOB-1293] Feature: Add Design Components

anadzam opened the pull request:
UI part of one screen chatting app 
1. Created one view, called ChatView which is instantiated twice in the VC, since we have two identical chat view on the screen.
2. Created separate component for switchButton which allows the user to set dark/light mode color.
3. Created separate component for textview, which is dynamically increases its height after five lines and is scrollable.
4. Added UIVIew for yellow line which divides top and bottom chatView.

UI supports both the landscape mode and any other device 

Logic part
1. Sent messages appear on both top and bottom tableview with appropriate sender cell or receiver cell.
2. Messages are saved in core data
3. When there is no network, the sent message only appears on sender tableview with fade design and error label
4. dark mode is saved in userDefaults.

Figma link: https://www.figma.com/file/UgTXNAXH2ScX6OHRREkSIZ/Designs-for-interns?type=design&node-id=0-1&t=IHbNQVRpFkbLvfLN-0


<img src="https://github.com/anadzam/Chat_App/assets/84043643/c95298a3-2f44-4cb6-8792-f43e55337c39" data-canonical-src="https://github.com/anadzam/Chat_App/assets/84043643/c95298a3-2f44-4cb6-8792-f43e55337c39" width="200" height="400"/>
<img src="https://github.com/anadzam/Chat_App/assets/84043643/a2669fc8-3790-47e5-813b-73474e17c9a6" data-canonical-src="https://github.com/anadzam/Chat_App/assets/84043643/a2669fc8-3790-47e5-813b-73474e17c9a6" width="200" height="400" />
<img src="https://github.com/anadzam/Chat_App/assets/84043643/46077f58-dd2b-4082-9f72-8dc4e4e76948" data-canonical-src="https://github.com/anadzam/Chat_App/assets/84043643/46077f58-dd2b-4082-9f72-8dc4e4e76948" width=400" height="200" />
<img src="https://github.com/anadzam/Chat_App/assets/84043643/5cf5daec-721e-4d82-82ac-c5999855bc68" data-canonical-src="https://github.com/anadzam/Chat_App/assets/84043643/5cf5daec-721e-4d82-82ac-c5999855bc68" width=400" height="200" />

anadzam committed:
added design components

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+
+class SwitchButton: UIButton {
+  var view = TopView()
+
```
view-ს არ იყენებ არსად ამიტომ წაშალე

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+  var view = TopView()
+
+    var status: Bool = false {
+        didSet {
```
`status` ის ნაცვლად `isOn` დაარქვი, უფრო ინტუიტიურია

LiziChichua replied:
გარედან წვდომადი არ უნდა იყოს ამიტომ `private` ქივორდი დაუწერე

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+    var onImage = UIImage(named: "lightMode")
+    var offImage = UIImage(named: "darkMode")
+    
```
იმიჯები კონსტანტებში გაიტანე. გააკეთე კონსტანტების სტრუქტურა ან ენამი. მე ენამს გირჩევ, სტრუქტურა ცოტა old-fashioned მიდგომაა

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    func update() {
+        UIView.transition(with: self, duration: 0.10, options: .transitionCrossDissolve, animations: {
```
გარედან წვდომადი არ უნდა იყოს ამიტომ `private` ქივორდი დაუწერე

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+        
+       
+    }
+    func toggle() {
```
გარედან წვდომადი არ უნდა იყოს ამიტომ `private` ქივორდი დაუწერე

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+   
+    
+    func setStatus(_ status: Bool) {
```
გარედან წვდომადი არ უნდა იყოს ამიტომ `private` ქივორდი დაუწერე

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+        self.toggle()
+    }
+    
+    func sendHapticFeedback() {
```
გარედან წვდომადი არ უნდა იყოს ამიტომ `private` ქივორდი დაუწერე

LiziChichua commented on Chat_App/Views/SwithcButton.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    
+  
```
ამ კლასში ძალიან ბევრი კომენტარი და extra lines გაქვს. წაშალე და მოაწესრიგე კოდი

LiziChichua commented on Chat_App/Views/TopView.swift:
```diff
@@ -0,0 +1,89 @@
+class TopView: UIView {
+    
+    
+    internal var sendButton: UIButton {
```
`internal` ის ნაცვლად `private` დაუწერე. ზოგადად დეფაულტად ისედაც `interna - ი` არის ამიტომ ნებისმიერ შემთხვევაში გინდა თუ არ გინდა რომ ინტერნალი იყოს ამ ქივორდის დაწერა საჭირო არ არის

LiziChichua commented on Chat_App/Views/TopView.swift:
```diff
@@ -0,0 +1,89 @@
+        let sendButton = UIButton(type: .custom)
+        sendButton.setImage(UIImage(named: "sendButton"), for: .normal)
+        sendButton.imageEdgeInsets = UIEdgeInsets(top: 0, left: -16, bottom: 0, right: 0)
+        sendButton.frame = CGRect(x: CGFloat(22.67), y: CGFloat(309.67), width: CGFloat(11.73), height: CGFloat(10.67))
```
კონსტრეინტების დასადებად გამოიყენე `NSLayoutConstraint`

LiziChichua commented on Chat_App/Views/TopView.swift:
```diff
@@ -0,0 +1,89 @@
+    //    @IBAction func refresh(_ sender: Any) {
+    //    }
+    
+    private let scrollView: UIScrollView = {
```
`scrollView` საერთოდ არ გჭირდება ამიტომ წაშალე. მესიჯებისთვის `UITableView-ს` გამოყენება მოგიწევს.

LiziChichua commented on Chat_App/Views/TopView.swift:
```diff
@@ -0,0 +1,89 @@
+        textField.layer.borderColor = UIColor.borderColor.cgColor
+        textField.isUserInteractionEnabled = true
+        
+        let myTextField = UITextField(frame: CGRect(x: 0, y: 0, width: 343, height: 56))
```
კონსტრეინტების დასადებად გამოიყენე `NSLayoutConstraint`

LiziChichua commented on Chat_App/Views/BottomView.swift:
```diff
@@ -0,0 +1,84 @@
+import UIKit
+
+class BottomView: UIView {
+    
```
ეს კლასი საერთოდ უნდა წაშალო, რადგან ზუსტად იგივე კოდი გიწერია რაც `TopView - ში`. ეს ორი ვიუ არის იდეურად აბსოლუტურად ერთი და იგივე ამიტომ დატოვე ერთი კლასი და სახელი გადაარქვი (Top/Bottom აღარ ერქვას) მაგ: ChatView. შემდეგ ამ ChatView - ს გამოიყენებ ორივე შემთხვევაში

LiziChichua commented on Chat_App/Resources/Extensions.swift:
```diff
@@ -0,0 +1,52 @@
+
+//extension UIColor {
+//    static let placeholderColor = UIColor(red: 199.0, green: 199.0, blue: 199.0, alpha: 1.0)
+//    static let borderColor = UIColor(red: 159.0, green: 96.0, blue: 225.0, alpha: 1.0)
```
ფერები ყოველთვის კონსტანტებში გაიტანე

LiziChichua commented on Chat_App/Resources/Extensions.swift:
```diff
@@ -0,0 +1,52 @@
+//}
+
+
+extension UIColor {
```
ექსთენშენებისთვის სხვადასხვა ფაილი გააკეთე. მაგ: UIColor+Extensions, UIView+Extensions და ა.შ

LiziChichua commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,100 @@
+
+class ViewController: UIViewController, UITextFieldDelegate {
+    
+    var textField = TopView()
```
უფრო ინტუიტიური სახელები დაარქვი... მაგ: topChatView/bottomChatView

LiziChichua replied:
საერთოდ `textField` ის კოდით დამატება არ გჭირდება. როგორც ვნახე სთორიბორდზე გაქვს `TopView`

LiziChichua commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    
+    
+    
+    private let scrollView: UIScrollView = {
```
არ გჭირდება `scrollView`. ეს არის კონტროლერი, კონტროლერში არც scrollView გჭირდება არც TableView იმიტომ, რომ მთლიანი სქრინი არ უნდა ისქროლებოდეს - უნდა ისქროლებოდეს მხოლოდ ChatView ში არსებული ჩატი (ანუ ტექსტფილდის ზევით რა სივრცეც არის ეგ)

LiziChichua commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    
+    
+    @objc func didToggleButton(_ sender: SwitchButton) {
+        if sender.status == true {
```
view.backgroundColor = sender.status ? .systemPink : .white

LiziChichua reviewed (changes requested):
1 - PR-ის გაკეთების დროს description-ის დაწერა არ დაგავიწყდეს + ჯირას ლინკიც მიუთითე
2 - Rectangle იმიჯი წაშალე, ამის ნაცვლად UIView გამოიყენე
3 - SwithcButton.swift ამ ფაილის სახელში ე.წ typo გაქვს - SwitchButton.swift უნდა იყოს
4 - ყველა ფაილში cmd + A & control + i გააკეთე ტაბების გასასწორებლად
5 - კონსტრეინტების დასადებად გამოიყენე NSLayoutConstraint
6 - ზედმეტი ხაზები წაშალე
7 - პატარა დევაისზე UI ინგრევა… UIStackView გამოიყენე

<img src="https://user-images.githubusercontent.com/83401753/233687590-e7f1c39e-900b-45fc-9a13-6f942debaf11.png" data-canonical-src="https://user-images.githubusercontent.com/83401753/233687590-e7f1c39e-900b-45fc-9a13-6f942debaf11.png" width="200" height="400" />

anadzam committed:
Added new components

Added tableView, textView and sendButton inside stackView. Added functionality of displaying textView inputs on tableView cell.

anadzam committed:
Fixed constraints

1.Removed components from storyboard.
2.Changed all CGrect with NSLayoutConstraints
3. corrected constraints to fix landscape issue

anadzam committed:
fixed PR issues

1. removed components from storyboard
2.fixed constraints issue to work on any device and landscape
3.changed all CGrect with NSLAyoutConstraints
4. added darkMode functionality to switch button

anadzam committed:
Defined global constants

Organized constants for colors, assets, placeholder... removed magic numbers

anadzam committed:
Added scrollable textview

Textview and its containerView is increasing height after 5 lines

shineRR commented on Chat_App/Assets.xcassets/rectangle.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "rectangle.png",
```
Use pdf files

shineRR commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,85 @@
+    private lazy var topChatView = ChatView()
+    private lazy var bottomChatView = ChatView()
+    
+    //MARK: - Yellow center line
```
I don't quite understand the purpose of using pragma marks as documentation for properties/methods.
https://nshipster.com/pragma/

shineRR commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,85 @@
+import UIKit
+
+
+class ViewController: UIViewController, UITextFieldDelegate {
```
_UITextFieldDelegate_ is redundant here, isn't it?

shineRR commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,85 @@
+
+class ViewController: UIViewController, UITextFieldDelegate {
+    
+    private lazy var switchButton = SwitchButton()
```
private let?

shineRR commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,85 @@
+    private lazy var bottomChatView = ChatView()
+    
+    //MARK: - Yellow center line
+    let centerLine: UIView = {
```
private?

shineRR commented on Chat_App/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,85 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        switchButton.addTarget(self, action: #selector(didTapButton), for: .touchUpInside)
```
i'd suggest to split this code into methods:
```
setupLayout()
setupConstraints()
setupActions/setupBindings()
```

shineRR commented on Chat_App/Extensions/UIColor+Extensions.swift:
```diff
@@ -0,0 +1,18 @@
+    convenience init(r: CGFloat, g: CGFloat, b: CGFloat) {
+        self.init(red: r/255, green: g/255, blue: b/255, alpha: 1)
+    }
+    
```
```suggestion

```

shineRR replied:
remove extra spaces

shineRR commented on Chat_App/Extensions/UIView+Extensions.swift:
```diff
@@ -0,0 +1,38 @@
+
+extension UIView {
+    
+    public var width: CGFloat {
```
why are properties `public`?

shineRR commented on Chat_App/Extensions/UIView+Extensions.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+    
+    public var height: CGFloat {
+        return self.frame.size.height
```
get rid of `self`

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+import UIKit
+import CoreData
+
+class ChatView: UIView {
```
final?

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+
+class ChatView: UIView {
+    let tableView = UITableView()
+    var sentMessages = [Message]()
```
not the best practice to store messages inside view

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+import CoreData
+
+class ChatView: UIView {
+    let tableView = UITableView()
```
private for properties?

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+        //        commented since we have that argument inside the function
+        //        let request: NSFetchRequest<Item> = Item.fetchRequest()
+        do {
+            sentMessages = try context.fetch(request)
```
view shouldn't fetch messages. Create a viewModel or any object that will be responsible for it

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+        tableView.backgroundColor = .clear
+        tableView.separatorStyle = .none
+        tableView.dataSource = self
+        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
```
Move identifier to `Constants` extension

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+        fatalError("")
+    }
+    
+    private func componentArranger() {
```
This name is like a class name, it's better to rename it

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+        
+        cell.textLabel?.textColor = Constants.Colors.textColor
```
I suppose it's better to move the UI configuration code inside the cell

shineRR commented on Chat_App/Views/ChatView.swift:
```diff
@@ -0,0 +1,154 @@
+        return cell
+    }
+    
+    
```
remove extra spaces

shineRR commented on Chat_App/Views/LeftBubble.swift:
```diff
@@ -0,0 +1,36 @@
+//    dateFormatter.dateFormat = "MM-dd HH:mm:ss"
+//    let dateString = dateFormatter.string(from: date)
+
+class LeftBubble: UIView {
```
final?

shineRR commented on Chat_App/Views/LeftBubble.swift:
```diff
@@ -0,0 +1,36 @@
+import Foundation
+import UIKit
+
+//    let date = Date()
```
remove comments

shineRR commented on Chat_App/Views/RightBubble.swift:
```diff
@@ -0,0 +1,39 @@
+
+class RightBubble: UIView {
+    
+    //    let date = Date()
```
the same comment? Please create extension and move this code to extension

shineRR commented on Chat_App/Views/RightBubble.swift:
```diff
@@ -0,0 +1,39 @@
+        Constants.Colors.bubbleColor.setFill()
+        secondBubble.fill()
+    }
+    
```
remove extra spaces

shineRR commented on Chat_App/Views/SwitchButton.swift:
```diff
@@ -0,0 +1,64 @@
+import Foundation
+import UIKit
+
+class SwitchButton: UIButton {
```
final?

shineRR commented on Chat_App/Views/SwitchButton.swift:
```diff
@@ -0,0 +1,64 @@
+    
+    var isOn: Bool = false {
+        didSet {
+            self.update()
```
get rid of `self`

shineRR commented on Chat_App/Views/SwitchButton.swift:
```diff
@@ -0,0 +1,64 @@
+    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
+        super.touchesEnded(touches, with: event)
+        self.sendHapticFeedback()
+        self.toggle()
```
```suggestion
        isOn.toggle()
```

shineRR replied:
https://developer.apple.com/documentation/swift/bool/toggle()

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+
+class TextView: UIView {
+    
+    var lineHeight: CGFloat {
```
why is it a computed property?

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+import UIKit
+
+
+class TextView: UIView {
```
Doesn't `UITextView` suit your solution?
https://developer.apple.com/documentation/uikit/uitextview

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+    
+    public func setUpButton() {
+        sendButton.setImage(UIImage(named: Constants.AssetIdentifier.sendButton.rawValue), for: .normal)
+        //        sendButton.addTarget(self, action: #selector(sendMessage), for: .touchUpInside)
```
remove comment

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+    var lineHeight: CGFloat {
+        Constants.TextView.font.lineHeight
+    }
+    public lazy var sendButton = UIButton()
```
why is it public?

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+//MARK: - TextView delegate methods
+extension TextView: UITextViewDelegate {
+    func textViewDidBeginEditing(_ textView: UITextView) {
+        if textView.textColor == Constants.Colors.placeholderColor {
```
I would suggest to use guard to avoid unnecessary nesting

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+    func textViewDidChange(_ textView: UITextView) {
+        let maxNumberOfLines = 6
```
Move this to TextView's `Constants` extension

shineRR commented on Chat_App/Extensions/Constants.swift:
```diff
@@ -0,0 +1,81 @@
+import Foundation
+import UIKit
+
+enum Constants {
```
You can create a private extension for your views and keep all constants there. For instance:
`private extension TextView {
       enum Constants { ... }
} `
`private extension SendButton {
       enum Constants { ... }
} `

anadzam replied:
you mean to create separate private extension files for every view?

shineRR replied:
yes

shineRR commented on Chat_App/Views/TextView.swift:
```diff
@@ -0,0 +1,135 @@
+        let fixedWidth = textView.frame.size.width
+        let newSize = textView.sizeThatFits(CGSize(width: fixedWidth, height: .infinity))
+        let newHeight = min(newSize.height, CGFloat(maxNumberOfLines) * lineHeight)
+        textView.constraints.forEach { constraint in
```
I think it's expensive to iterate over all constraints. 
You can store your height constraint inside the view and directly assign a new value to it

shineRR commented on Chat_App/Extensions/UIImage+Extensions.swift:
```diff
@@ -0,0 +1,16 @@
+extension UIImage {
+    
+    convenience init(assetIdentifier: Constants.AssetIdentifier) {
+        self.init(named: assetIdentifier.rawValue)!
```
```suggestion
        self.init(named: assetIdentifier.rawValue)
```

shineRR replied:
What is the purpose of this `convenience init`?

shineRR commented:
Please fix `textColor` in Dark mode, text can't be read
<img width="276" alt="Screenshot 2023-05-15 at 09 32 53" src="https://github.com/anadzam/Chat_App/assets/38408304/27ba4e90-f975-47ed-b25c-1c617a50ceca">

anadzam committed:
added chat bubbles

anadzam committed:
added newtwork check

While sending message, network connection is checked and label below the messages is set based on the result.

anadzam committed:
added darkMode to userDefaults, fixed issues in switchbutton

anadzam committed:
modified send message function, fixed keyboard dismiss

anadzam committed:
Removed Storyboard, fixed issues

anadzam committed:
fixed faildToSend logic

sadnessReigns commented on Chat_App/Core/CoreData/CoreDataManager.swift:
```diff
@@ -0,0 +1,53 @@
+    let request: NSFetchRequest<MessageEntity> = MessageEntity.fetchRequest()
+    
+    //    MARK: - Load messages from coreData
+    func loadObjects<T: NSManagedObject>(with request: NSFetchRequest<T>) -> [T]? {
```
better would be making such functions as `throws` and create error system for CoreDataManager with enums and localized descriptions (or at least just descriptions) and throw those errors from these functions instead of `print` and `return nil`. Cause that way you'll be able to use it via `do-try-catch` and in `catch` block on the side you're calling these functions you'll be able to present corresponding error message on a screen (such as like "Sorry! We weren't be able to load your messages history!")

sadnessReigns commented on Chat_App/Core/Custom UI/ChatView/ChatView.swift:
```diff
@@ -0,0 +1,128 @@
+
+class ChatView: UIView {
+    
+    let tableView = UITableView()
```
are views accessed from the outside of that view? If yes, at least make then public private(set). better - create a separate API for that purpose (e.g. `func readTextViewTextChanges() -> String` and call such functions either at appropriate delegate (UITextViewDelegate in that case) or at appropriate view events

sadnessReigns commented on Chat_App/Core/Custom UI/ChatView/ChatView.swift:
```diff
@@ -0,0 +1,128 @@
+    var typingArea = TextView()
+    var viewModel = ChatViewModel()
+    weak var sendMessageDelegate: SendMessageDelegate?
+    var currentUser = 1 {
```
current user is Int? 
Me as your teammate require either a documentation of which numbers mean what or at least a comment.

sadnessReigns commented on Chat_App/Core/Custom UI/ChatView/ChatView.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+    
+    func configureTableView() {
+        tableView.translatesAutoresizingMaskIntoConstraints = false
```
this property is related to constraints. It would be better off inside method that builds constraints.

sadnessReigns replied:
also this method doesn't have to be accessed from the outside. Make it private or move to a private extension

sadnessReigns commented on Chat_App/Core/Custom UI/ChatView/ChatView.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    func setUpConstraints() {
+        NSLayoutConstraint.activate([
```
would be better to create a sequence of functions that build constraints individually for each subview and call then inside of `setUpConstraints `

sadnessReigns commented on Chat_App/Core/Custom UI/ChatView/ChatView.swift:
```diff
@@ -0,0 +1,128 @@
+extension ChatView: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        //        viewModel.numberOfMessages()
+        return viewModel.numberOfMessages(currentUser: currentUser)
```
omit `return`

sadnessReigns commented on Chat_App/Core/Custom UI/ChatView/ChatView.swift:
```diff
@@ -0,0 +1,128 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let message = viewModel.message(at: indexPath.row, currentUser: currentUser)
+        if message.userId == currentUser {
```
if `currentUser` field is current user's id then move it to viewModel. ViewController doesn't need to store it

sadnessReigns commented on Chat_App/Core/Custom UI/SwitchButton/SwitchButton.swift:
```diff
@@ -0,0 +1,50 @@
+
+class SwitchButton: UIButton {
+    
+    var isOn: Bool = false {
```
I would prefer it to be a public private(set) to avoid changing from the outside

sadnessReigns commented on Chat_App/Core/Custom UI/TableViewCells/ReceiverCell.swift:
```diff
@@ -0,0 +1,131 @@
+        contentView.addSubview(bigBubble)
+        contentView.addSubview(mediumBubble)
+        contentView.addSubview(smallBubble)
+        label.translatesAutoresizingMaskIntoConstraints = false
```
these are not a part of setting up hierarchy, these are part of constraints setup

sadnessReigns commented on Chat_App/Core/Custom UI/TableViewCells/ReceiverCell.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+    
+    private func setUpLayoutConstraints() {
+        NSLayoutConstraint.activate([
```
create individual methods for each view and call them here. Also if you wanna be a rockstar, call your methods consistently, on a previous file I saw `setUpConstraints` which is a better way of calling a method. Use it here as well

sadnessReigns commented on Chat_App/Core/Custom UI/TableViewCells/SenderCell.swift:
```diff
@@ -0,0 +1,134 @@
+//
+//  Created by Ana Dzamelashvili on 5/29/23.
+//
+
```
same comments as for ReceiverCell

sadnessReigns commented on Chat_App/Core/Custom UI/TextView/TextView.swift:
```diff
@@ -0,0 +1,160 @@
+    private let lineHeight: CGFloat = Constants.TextView.font.lineHeight
+    private lazy var sendButton = UIButton()
+    private lazy var containerView = UIView()
+    lazy var textView = UITextView()
```
make it public private(set), you probably don't need to set that textView from the outside

sadnessReigns commented on Chat_App/Core/Custom UI/TextView/TextView.swift:
```diff
@@ -0,0 +1,160 @@
+    }
+    
+    func getSendButton() -> UIButton {
+           return sendButton
```
omit return

sadnessReigns replied:
or alternatively you can make `sendButton`'s getter public and setter private (`public private(set)`) and get rid of that method

sadnessReigns commented on Chat_App/Core/Custom UI/TextView/TextView.swift:
```diff
@@ -0,0 +1,160 @@
+    }
+    //MARK: - Constraints
+    func setUpConstraints() {
+        
```
create individual methods for constraints

sadnessReigns commented on Chat_App/Core/Custom UI/TextView/TextView.swift:
```diff
@@ -0,0 +1,160 @@
+        let fixedWidth = textView.frame.size.width
+        let newSize = textView.sizeThatFits(CGSize(width: fixedWidth, height: .infinity))
+        let newHeight = min(newSize.height, Constants.TextView.maxNumberOfLines * lineHeight)
+        textView.constraints.forEach { constraint in
```
this is pretty invasive method of changing constraint's properties. Better way would be to store the height constraint like:
private lazy var textViewHeightConstraint = textView.heightAnchor.constraint(....)

then activate it somewhere in your setup methods. Then change it's constant right here and `UIView.animate` with `layoutIfNeeded()` call inside if you need to animate these height changes

sadnessReigns commented on Chat_App/Core/Custom UI/TextView/TextView.swift:
```diff
@@ -0,0 +1,160 @@
+            }
+        }
+        
+        containerView.constraints.forEach { constraint in
```
same right there

sadnessReigns commented on Chat_App/Core/Custom UI/ViewModel/ChatViewModel.swift:
```diff
@@ -0,0 +1,88 @@
+    private lazy var sentMessages = [MessageEntity]()
+  
+    
+    func filteredMessages(currentUser: Int) -> [MessageEntity] {
```
see? If you move currentUser to viewModel you can even omit a function parameter right there and since your current user will be stored right there inside viewModel you'll be able to call this function without a parameter at all, but renaming would be a good idea as well. Something like `filteredMessagesForCurrentUser()` would be a decent fit, but you can come up with your own naming

sadnessReigns commented on Chat_App/Core/Custom UI/ViewModel/ChatViewModel.swift:
```diff
@@ -0,0 +1,88 @@
+    
+    //MARK: - Load messages from coreData
+    func loadMessages() {
+        if let messages = coreDataManager.loadObjects(with: coreDataManager.request) {
```
see what I was talking about in CoreDataManager? 
If `loadObjects(with:)` would be a method that throws, you could be doing there just 
do {
sendMessages = try coreDataManager.loadObjects(with:)
delegate?.messagesLoaded()
} catch error as? CoreDataManager.InternalError { // syntax is pseudo, but you can check documentation on that
delegate.sendError(error)
}

sadnessReigns replied:
and no optionals unwrapping while working with your manager. You'd be able to throw errors at your delegate that would be showing certain views in that case or maybe an UIAlertController

sadnessReigns commented on Chat_App/Core/Custom UI/ViewModel/ChatViewModel.swift:
```diff
@@ -0,0 +1,88 @@
+//        return sentMessages.count
+//    }
+    
+    func numberOfMessages(currentUser: Int) -> Int {
```
again if currentUser would be in viewModel that function would be without a parameter

sadnessReigns commented on Chat_App/Core/Extensions/UIColor+Extensions.swift:
```diff
@@ -0,0 +1,27 @@
+
+extension UIColor {
+    convenience init(red: Int, green: Int, blue: Int) {
+        assert(red >= 0 && red <= 255, "Invalid red component")
```
a rockstar badge for you if you came up with that by yourself. Defending your code from dumbasses is always a good practice

sadnessReigns commented on Chat_App/Scenes/Controller/ViewController.swift:
```diff
@@ -0,0 +1,181 @@
+    }
+    
+    //MARK: - Set up constraints
+    private func constraintsAssigner() {
```
naming issue there: `constraintsAssigner` is an entity name. You should name methods by verbs and describe what they do

sadnessReigns commented on Chat_App/Scenes/Controller/ViewController.swift:
```diff
@@ -0,0 +1,181 @@
+    //MARK: - Set up constraints
+    private func constraintsAssigner() {
+        
+        NSLayoutConstraint.activate([
```
also create individual methods for setting up constraints for each view

anadzam committed:
fixed issues based on PR comments

anadzam committed:
removed viewmodel instance from Chatview Class

anadzam committed:
retrieved network logic
